python/ray/data/_internal/planner/plan_udf_map_op.py

```diff
@@ -606,3 +608,194 @@ def _create_map_transformer_for_block_based_map_op(
+                        cur_task_map[task] = idx
+                    except StopIteration:
+                        consumed = True
+                        break
+
+                # Check if any running tasks remaining
+                if not cur_task_map:
+                    break
+
+                done, pending = await asyncio.wait(
+                    cur_task_map.keys(), return_when=asyncio.FIRST_COMPLETED
+                )
+
+                for task in done:
+                    # Report completed tasks along w/ its corresponding
+                    # index in the input sequence
+                    #
+                    # NOTE: Once completed tasks queue fills up, this will block
+                    #       therefore serving as back-pressure for scheduling tasks
+                    #       preventing it from scheduling new tasks
+                    await completed_tasks_queue.put((task, cur_task_map[task]))
+
+                    cur_task_map.pop(task)
+
+        except BaseException as e:
+            for cur_task in cur_task_map:
+                if not cur_task.done():
+                    cur_task.cancel()
+
+            sentinel = e
+        finally:
+            assert len(cur_task_map) == 0, f"{cur_task_map}"
+            await completed_tasks_queue.put((sentinel, None))
+
+    async def _consume(
+        completed_tasks_queue: asyncio.Queue, output_queue: queue.Queue
+    ) -> None:
+        completed_task_map: Dict[int, asyncio.Task] = dict()
+        next_idx = 0
+        completed_scheduling = False
+
+        try:
+            while not completed_scheduling:
+                task, idx = await completed_tasks_queue.get()
+
+                if isinstance(task, Exception):
+                    raise task
+                elif task is _SENTINEL:
+                    completed_scheduling = True
+                else:
+                    completed_task_map[idx] = task
+
+                while next_idx in completed_task_map:
+                    next_task = completed_task_map.pop(next_idx)
+
+                    # NOTE: Once output queue fills up, this will block
+                    #       therefore serving as back-pressure for scheduling tasks
+                    #       preventing it from scheduling new tasks.
+                    # NOTE: This will block the whole event-loop not just this task
+                    output_queue.put(await next_task)
+
+                    next_idx += 1
+
+            assert (
+                len(completed_task_map) == 0
+            ), f"{next_idx=}, {completed_task_map.keys()=}"
+            sentinel = _SENTINEL
+
+        except BaseException as e:
+            sentinel = e
+        finally:
+            output_queue.put(sentinel)
+
+    def _transform(batch_iter: Iterable[T], task_context: TaskContext) -> Iterable[U]:
+        task_completion_queue = asyncio.Queue(maxsize=max_concurrency)
+        output_queue = queue.Queue(maxsize=max_concurrency)
```

Comment:
With this, the actual concurrency will be `max_concurrency ** 2`. 
Because when `max_concurrency` is set on a map, the actor will be running `max_concurrency` threads.
And in each thread, we allow `max_concurrency` asyncio tasks.
We should make these queues global to all threads (put them in `_map_actor_context`. 

Note, the more proper fix is to just make the Actor an async actor, so Ray Core can natively handle the concurency.

Reply:
> With this, the actual concurrency will be max_concurrency ** 2.

That's not true -- async actors provision just 1 event-loop (it's actually 1 event-loop per concurrecny group, but certainly not equal to the max_concurrency)

BTW, any actor with async methods is already considererd async by Core

Reply:
Discussed offline:

 - UDF isn't an actor, but just part of its (_MapWorker) state
 - There's just 1 event-loop per Actor
 - We're going to use configuration (defaults to 4) separate from Core's `max_concurrency`